File: vs2003_cd01/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/Samples/Technologies/Interop/Applications/ComServices/VC7Bank/GetReceipt.cpp
// ==============================================================================
// Filename: GetReceipt.h
//
// Summary:  managed cpp definition of the GetReceipt class for the bank sample
// Classes:  GetReceipt.cs
//
// This file is part of the Microsoft COM+ Samples
//
// Copyright (C) Microsoft Corporation. All rights reserved
//
// This source code is intended only as a supplement to Microsoft
// Development Tools and/or on-line documentation.  See these other
// materials for detailed information reagrding Microsoft code samples.
//
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
// KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
// PARTICULAR PURPOSE.
//


// import mscorlib definitions
#using "mscorlib.dll"
using namespace System;

// import definitions for EnterpriseServices support
#using "System.EnterpriseServices.dll"
using namespace System::EnterpriseServices;
using namespace System::Runtime::InteropServices;


//import definitions for ADO
#using "adodb.dll"
using namespace ADODB;

// import definition for account typelibrary
#using "AccountComLib.dll"
using namespace AccountComLib;

#define null 0

#include "getreceipt.h"
#include "updatereceipt.h"

namespace VC7Bank
{
    int GetReceipt::GetNextReceipt ()
    {

        bool bSuccess = false;
        int result;

        // First of all, get the object context
        try
        {
            // Call the true function
            result = trueGetNextReceipt ();
            bSuccess = true;
            return result;
        }

        // Upon exit, always call SetComplete if happy, or SetAbort if unhappy.
        // We do this because we never save state across method calls.
        __finally
        {
            if (bSuccess)
                ContextUtil::SetComplete();
            else
                ContextUtil::SetAbort();
        }
    }


    // F+F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++
    //
    // Function: trueGetNextReceipt
    //
    // trueGetNextReceipt() is the function that performs the actual work for the Account class.
    // If an error occurs during execution, it will throw a COMException for GetNextReceipt()
    // to handle.
    //
    // Args:     None
    // Returns:  Long -  next receipt value
    //
    // F-F---F---F---F---F---F---F---F---F---F---F---F---F---F---F---F---F---F---F---


    // trueGetNextReceipt using COM+ Shared Properties:
    int GetReceipt::trueGetNextReceipt ()
    {
        SharedPropertyGroupManager* spmMgr = null;
        SharedPropertyGroup* spmGroup = null;
        SharedProperty* spmPropNextReceipt = null;
        SharedProperty* spmPropMaxNum = null;
        IUpdateReceipt* objReceiptUpdate = null;

        try
        {
            // Create SPM group manager
            spmMgr = new SharedPropertyGroupManager();

            bool bExists = false;
            PropertyLockMode iLockMode;  //= new int [1];
            PropertyReleaseMode iReleaseMode;//  = new int [1];

            //iLockMode [0] = ISharedPropertyGroupManager.LOCKMODE_METHOD;
            iLockMode  = PropertyLockMode::Method;
            //iReleaseMode [0] = ISharedPropertyGroupManager.RELEASEMODE_PROCESS;
            iReleaseMode  = PropertyReleaseMode::Process;

            // Create the shared property group
            spmGroup = spmMgr->CreatePropertyGroup (S"Receipt", &iLockMode, &iReleaseMode, &bExists);

            // Create the properties
            spmPropNextReceipt = spmGroup->CreateProperty(S"Next", &bExists);
            spmPropMaxNum = spmGroup->CreateProperty(S"MaxNum", &bExists);

            // Call GetNextReceipt() if necessary
            int iPropNextReceipt = *dynamic_cast<Int32*>(spmPropNextReceipt->Value);
            int iPropMaxNum = *dynamic_cast<Int32*>(spmPropMaxNum->Value); 
            if( iPropNextReceipt >= iPropMaxNum)
            {

				objReceiptUpdate = dynamic_cast<IUpdateReceipt __gc *>(new UpdateReceipt());
                int iRet = objReceiptUpdate->Update ();
                spmPropNextReceipt->Value = __box(iRet);
                spmPropMaxNum->Value = __box(iRet + 100);
            }

            int vfoo = *dynamic_cast<Int32*>(spmPropNextReceipt->Value) + 1;
            spmPropNextReceipt->Value = __box(vfoo);

            // We are finished and happy
            return (int)*dynamic_cast<Int32*>(spmPropNextReceipt->Value);
        }

        __finally
        {
        }
    }
}

File: vs2003_cd01/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/Samples/Technologies/Interop/Applications/ComServices/VB7Bank/GetReceipt.vb

' ==============================================================================
' Filename: Account.vbp
'
' Summary:  Visual Basic implememtation of the GetReceipt component for the bank sample
' Classes:  GetReceipt.cls
'
' This file is part of the Microsoft COM+ Samples
'
' Copyright (C) Microsoft Corporation. All rights reserved
'
' This source code is intended only as a supplement to Microsoft
' Development Tools and/or on-line documentation.  See these other
' materials for detailed information reagrding Microsoft code samples.
'
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
' KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
' IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
' PARTICULAR PURPOSE.
'
'

option strict off

imports System
imports ADODB
imports AccountComLib
imports System.EnterpriseServices
imports Microsoft.VisualBasic
imports System.Runtime.InteropServices

namespace VB7Bank

	<TransactionAttribute(TransactionOption.Supported)> public class GetReceipt		
			inherits   ServicedComponent
			implements IGetReceipt

		Private Const APP_ERROR = -2147467008

		' F+F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++F+++
		'
		' Function: GetNextReceipt
		'
		' This function uses Shared Properties to get the next receipt value
		'
		' Args:     None
		' Returns:  Long -  next receipt value
		'
		' F-F---F---F---F---F---F---F---F---F---F---F---F---F---F---F---F---F---F---F---
		Public Function GetNextReceipt() As Integer Implements IGetReceipt.GetNextReceipt

			dim success as boolean
	
			success = false
			try
				dim spmMgr as new SharedPropertyGroupManager
				dim spmGroup as SharedPropertyGroup
				dim bResult as Boolean
				spmGroup = spmMgr.CreatePropertyGroup("Receipt", PropertyLockMode.Method, PropertyReleaseMode.Process, bResult)

				dim spmPropNextReceipt as SharedProperty
				spmPropNextReceipt = spmGroup.CreateProperty("Next", bResult)

				'  the initial value of the Shared Property to
				' 0 if the Shared Property didn’t already exist.
				' This is not entirely necessary but demonstrates how to initialize a value.
				if bResult = false then
					spmPropNextReceipt.Value = (0)
				end if

				dim spmPropMaxNum as SharedProperty
				spmPropMaxNum = spmGroup.CreateProperty("MaxNum", bResult)

				dim objReceiptUpdate as new VB7Bank.UpdateReceipt
				if spmPropNextReceipt.Value >= spmPropMaxNum.Value then
					spmPropNextReceipt.Value = ( objReceiptUpdate.Update)
					spmPropMaxNum.Value = (spmPropNextReceipt.Value + 100)
				end if

				' Get the next receipt number and update property
				spmPropNextReceipt.Value = (spmPropNextReceipt.Value + 1)

				GetNextReceipt = cint(spmPropNextReceipt.Value)
				success = true

			catch e as Exception
				GetNextReceipt = -1							' indicate that an error occured
				success = false
				throw new Exception("VB7Bank.GetReceipt.GetNextReceipt " & e.ToString())
			
			finally
				if success then
					ContextUtil.SetComplete()
				else
					ContextUtil.SetAbort()
				end if

			end try

		end function


	end class

end namespace
